app/registries/foreman/plugin/medium_providers_registry.rb

```diff
@@ -0,0 +1,47 @@
+module Foreman
+  class Plugin
+    class MediumProvidersRegistry
+      attr_reader :providers
+
+      def initialize
+        @providers = []
+      end
+
+      def register(provider)
+        provider = provider.to_s
+        @providers.unshift(provider)
+        provider
+      end
+
+      def find_provider(entity)
+        valid_provider = nil
+        errors = {}
+        providers.each do |provider_class|
+          provider = provider_class.constantize.new(entity)
+          if provider.valid?
+            valid_provider = provider
+            break
+          end
+
+          errors[provider.class.friendly_name] = provider.errors
+        end
+
+        unless valid_provider
+          raise Foreman::Exception.new 'Could not find a provider for %{entity}. Providers returned %{errors}', { entity: entity, errors: errors } if errors.present?
+        end
+
+        valid_provider
+      end
```

Comment:
I see one major problem here - the order. We cannot assume any order, we've seen bugs when plugin was renamed (discovery to foreman_discovery) which changed initialization order and broke things.

We can however make an assumption that only one provider must be valid, therefore is multiple are found we explicitly error out with an exception, this will prevent plugins from sneaking in providers which are pseudo-valid.